File: Cosmos-IOS/Cosmostation/Cosmostation/Model/KavaModel/CdpParam.swift
//
//  CdpParam.swift
//  Cosmostation
//
//  Created by 정용주 on 2021/03/04.
//  Copyright © 2021 wannabit. All rights reserved.
//

import Foundation

public struct CdpParam {
    var surplus_auction_threshold: String?
    var surplus_auction_lot: String?
    var debt_auction_threshold: String?
    var debt_auction_lot: String?
    var savings_distribution_frequency: String?
    var circuit_breaker: Bool?
    var global_debt_limit: Coin?
    var debt_param: DebtParam?
    var collateral_params: Array<CollateralParam>?
    
    init(_ dictionary: NSDictionary?) {
        self.surplus_auction_threshold = dictionary?["surplus_auction_threshold"] as? String
        self.surplus_auction_lot = dictionary?["surplus_auction_lot"] as? String
        self.debt_auction_threshold = dictionary?["debt_auction_threshold"] as? String
        self.debt_auction_lot = dictionary?["debt_auction_lot"] as? String
        self.savings_distribution_frequency = dictionary?["savings_distribution_frequency"] as? String
        self.circuit_breaker = dictionary?["circuit_breaker"] as? Bool
        if let rawGlobalDebtLimit = dictionary?["global_debt_limit"] as? NSDictionary {
            self.global_debt_limit = Coin.init(rawGlobalDebtLimit)
        }
        if let rawDebtParam = dictionary?["debt_param"] as? NSDictionary {
            self.debt_param = DebtParam.init(rawDebtParam)
        }
        if let rawCollateralParams = dictionary?["collateral_params"] as? Array<NSDictionary> {
            self.collateral_params = Array<CollateralParam>()
            for rawCollateralParam in rawCollateralParams {
                self.collateral_params?.append(CollateralParam.init(rawCollateralParam))
            }
        }
    }
    
    public func getCollateralParamByDenom(_ denom: String) -> CollateralParam? {
        return collateral_params?.filter { $0.denom == denom}.first
//        for param in collateral_params {
//            if (param.denom == denom) {
//                return param
//            }
//        }
//        return nil
    }
    
    public func getCollateralParamByType(_ type: String) -> CollateralParam? {
        return collateral_params?.filter { $0.type == type}.first
//        for param in collateral_params {
//            if (param.type == type) {
//                return param
//            }
//        }
//        return nil
    }
    
    public func getGlobalDebtAmount() -> NSDecimalNumber {
        var result = NSDecimalNumber.zero
        if (global_debt_limit != nil && !global_debt_limit!.denom.isEmpty) {
            result =  NSDecimalNumber.init(string: global_debt_limit?.amount)
        }
        return result
    }
    
}

File: app/src/main/java/wannabit/io/cosmostaion/model/kava/CdpParam.java
package wannabit.io.cosmostaion.model.kava;

import com.google.gson.annotations.SerializedName;

import java.math.BigDecimal;
import java.util.ArrayList;

import wannabit.io.cosmostaion.model.type.Coin;

public class CdpParam {
    @SerializedName("surplus_auction_threshold")
    public String surplus_auction_threshold;

    @SerializedName("surplus_auction_lot")
    public String surplus_auction_lot;

    @SerializedName("debt_auction_threshold")
    public String debt_auction_threshold;

    @SerializedName("debt_auction_lot")
    public String debt_auction_lot;

    @SerializedName("savings_distribution_frequency")
    public String savings_distribution_frequency;

    @SerializedName("circuit_breaker")
    public Boolean circuit_breaker;

    @SerializedName("global_debt_limit")
    public Coin global_debt_limit;

    @SerializedName("debt_param")
    public DebtParam debt_param;

    @SerializedName("collateral_params")
    public ArrayList<CollateralParam> collateral_params;

    public BigDecimal getGlobalDebtAmount() {
        BigDecimal result = BigDecimal.ZERO;
        if (global_debt_limit != null) {
            result = new BigDecimal(global_debt_limit.amount);
        }
        return result;
    }

    public BigDecimal getRawLiquidationRatio(String denom) {
        for (CollateralParam param : collateral_params) {
            if (param.denom.equals(denom)) {
                return new BigDecimal(param.liquidation_ratio);
            }
        }
        return BigDecimal.ZERO;
    }

    //only using for price
    public CollateralParam getCollateralParamByDenom(String denom) {
        CollateralParam result = null;
        for (CollateralParam collateralParam:collateral_params) {
            if (collateralParam.denom.equals(denom)) {
                return collateralParam;
            }
        }
        return  result;
    }


    public CollateralParam getCollateralParamByType(String type) {
        CollateralParam result = null;
        for (CollateralParam collateralParam:collateral_params) {
            if (collateralParam.type.equals(type)) {
                return collateralParam;
            }
        }
        return result;
    }

}
